From e27c7b4d1a5042c2498ae000117367081adea4f6 Mon Sep 17 00:00:00 2001
From: Joshua Strobl <joshua@streambits.io>
Date: Thu, 16 Jan 2020 10:44:22 +0200
Subject: [PATCH 1/1] allow confinement of AVX specific library paths

Since glibc 2.26 library lookups will now look first for specific AVX2 and AVX512 library versions on compatible hardware. This patch adds the directories to snap-confine, allowing snap to support distributions that ship these libraries.

Before: /usr/lib/snapd/snap-confine: error while loading shared libraries: libc.so.6: failed to map segment from shared object
After: Application loads as expected
---
 cmd/snap-confine/snap-confine.apparmor.in | 46 +++++++++++------------
 1 file changed, 22 insertions(+), 24 deletions(-)

diff --git a/cmd/snap-confine/snap-confine.apparmor.in b/cmd/snap-confine/snap-confine.apparmor.in
index d693df3..743d5bc 100644
--- a/cmd/snap-confine/snap-confine.apparmor.in
+++ b/cmd/snap-confine/snap-confine.apparmor.in
@@ -14,32 +14,30 @@
     # any abstractions
     /etc/ld.so.cache r,
     /etc/ld.so.preload r,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}ld-*.so mrix,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}ld-*.so mrix,
     # libc, you are funny
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libc{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libpthread{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libreadline{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}librt{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libgcc_s.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libncursesw{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libresolv{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libselinux.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libpcre.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libmount.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libblkid.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libuuid.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libc{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libpthread{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libreadline{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}librt{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libgcc_s.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libncursesw{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libresolv{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libselinux.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libpcre.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libmount.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libblkid.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libuuid.so* mr,
     # normal libs in order
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libapparmor.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libcgmanager.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libdl{,-[0-9]*}.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libnih.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libnih-dbus.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libdbus-1.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libudev.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libseccomp.so* mr,
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libcap.so* mr,
-    # Needed to run /usr/bin/sh for snap-device-helper.
-    /{,usr/}lib{,32,64,x32}/{,@{multiarch}/}libtinfo.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libapparmor.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libcgmanager.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libdl{,-[0-9]*}.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libnih.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libnih-dbus.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libdbus-1.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libudev.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libseccomp.so* mr,
+    /{,usr/}lib{,32,64,x32}/{,haswell/{,avx512_1/},@{multiarch}/}libcap.so* mr,
 
     @LIBEXECDIR@/snap-confine mr,
 
-- 
2.25.0

